File: searchnsort/ltcd_786.cpp
//LeetCode - problem 786

vector kthSmallestPrimeFraction(vector& nums, int k) {
    double low=0,high=1;
    int n=nums.size();
    while(low<=high){
        double mid=(low+high)/2;
        double maxf=0.0;
        int p,q;
        int total=0;
        int j=1;
        for(int i=0;i<n-1;i++){
            while(j<n && (double)nums[i]/(double)nums[j]>mid){
                j++;
            }
            total+=(n-j);
            if(j==n){
                break;
            }
            double frac=(double)nums[i]/(double)nums[j];
            if(frac>=maxf){
                maxf=frac;
                p=i;
                q=j;
                
            }
        }
        if(total==k){
            return {nums[p],nums[q]};
        }
        else if(total>k){
            high=mid;
        }
        else{
            low=mid;
        }
        
    }
    return {};
}

File: searchnsort/ltcd_786.py
'''
LeetCode - problem 786
'''

    def kthSmallestPrimeFraction(self, arr: List[int], k: int) -> List[int]:
        left, right = 0, 1
        p, q = 0, 1
        n = len(arr)
        while True:
            cnt = 0
            p = 0
            m = (left+right)/2
            for i in range(n):
                j = n - 1 - i
                while j >= 0 and arr[i] > m*arr[n-1-j]:
                    j -= 1
                cnt += j+1
                if j >= 0 and q*arr[i] > p*arr[n-1-j]:
                    p = arr[i]
                    q = arr[n-1-j]
            if cnt < k:
                left = m
            elif cnt > k:
                right = m
            else:
                return [p, q]
